fn.go

```diff
@@ -20,34 +32,204 @@ type Function struct {
+			if err == io.EOF {
+				break
+			}
+			response.Fatal(rsp, errors.Wrap(err, "cannot decode manifest"))
+			return rsp, nil
+		}
+		if u != nil {
+			objs = append(objs, u)
+		}
+	}
+
+	// Get the desired composite resource from the request.
+	dxr, err := request.GetDesiredCompositeResource(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot get desired composite resource"))
+		return rsp, nil
+	}
+
+	//  Get the desired composed resources from the request.
+	dcd, err := request.GetDesiredComposedResources(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot get desired composed resources"))
+		return rsp, nil
+	}
+
+	// Convert the rendered manifests to a list of desired composed resources.
+	for _, obj := range objs {
+		cd := resource.NewDesiredComposed()
+		cd.Resource.Unstructured = *obj.DeepCopy()
+
+		// Update only the status of the desired composite resource.
+		if cd.Resource.GetAPIVersion() == dxr.Resource.GetAPIVersion() && cd.Resource.GetKind() == dxr.Resource.GetKind() {
+			dst := make(map[string]any)
+			if err := dxr.Resource.GetValueInto("status", &dst); err != nil && !fieldpath.IsNotFound(err) {
+				response.Fatal(rsp, errors.Wrap(err, "cannot get desired composite status"))
+				return rsp, nil
+			}
+
+			src := make(map[string]any)
+			if err := cd.Resource.GetValueInto("status", &src); err != nil && !fieldpath.IsNotFound(err) {
+				response.Fatal(rsp, errors.Wrap(err, "cannot get desired composite status"))
+				return rsp, nil
+			}
+
+			for k, v := range src {
+				fmt.Println(k, v)
+				dst[k] = v
+			}
+
+			if err := fieldpath.Pave(dxr.Resource.Object).SetValue("status", dst); err != nil {
+				response.Fatal(rsp, errors.Wrap(err, "cannot set desired composite status"))
+				return rsp, nil
+			}
+
+			continue
+		}
+
+		// Set composite resource's connection details.
+		if cd.Resource.GetAPIVersion() == metaApiVersion {
+			switch obj.GetKind() {
+			case "CompositeConnectionDetails":
+				con, _ := cd.Resource.GetStringObject("data")
+				for k, v := range con {
+					d, _ := base64.StdEncoding.DecodeString(v) //nolint:errcheck // k8s returns secret values encoded
+					dxr.ConnectionDetails[k] = d
+				}
+			default:
+				response.Fatal(rsp, fmt.Errorf(errFmtInvalidMetaType, obj.GetKind()))
+				return rsp, nil
+			}
+
+			continue
+		}
+
+		// Set ready state.
+		if v, found := cd.Resource.GetAnnotations()[annotationKeyReady]; found {
+			if v != string(resource.ReadyTrue) && v != string(resource.ReadyUnspecified) && v != string(resource.ReadyFalse) {
+				response.Fatal(rsp, fmt.Errorf(fmt.Sprintf(errFmtInvalidFunction, errFmtInvalidReadyValue), v))
+				return rsp, nil
+			}
+
+			cd.Ready = fn.Ready(v)
+
+			// remove meta annotation
+			ann := cd.Resource.GetAnnotations()
+			delete(ann, annotationKeyReady)
+			cd.Resource.SetAnnotations(ann)
+		}
+
+		// Add resource to the desired composed resources map.
+		name, err := getCompositionResourceName(obj)
```

Comment:
nit: logic is simple enough to inline here instead of introducing a new function.